design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,233 @@
+
+However, we may have a limit on the size of the URL and we will have an issue  if the certificate will grow in the future. To overcome this, we can split the  request. The first one will create the file. The following requests will populate it with the cert content. Then the last request will verify that the whole content has passed (with one of the methods described above). We can decide on the number  of requests dynamically according to the size of the cert.
+
+We have a security concern with this option. While the traffic is fully encrypted when using https, including the URL, [it's still possible that the URL will be visible in certain locations](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), while the request travels to its destination. It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.
+
+Thus, we will not use this option due to security concerns.
+
+#### Option 3 - Client sets its State once cert is injected
+This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code).
+
+In this case, we respond with 202 Accepted to the "inject-client-cert" request and start injecting the cert in a separate thread. The server then polls on the state of the authn-client container and waits until its State changes to some  agreed State (e.g "CertInjected"). This state will be set by the authn-client once it sees that the cert is installed.
+
+This option is not fully backwards-compatible and is quite complex so I wouldn't suggest to implement it.
+
+#### Option 4 - Send cert as part of bash script
+
+The issue with the STDIN that is not closing _may_ be an issue of the `tar` command that waits for more input and doesn't close the connection. Another option that we have is to change the command from `tar` to `sh` and send a bash script in the STDIN.
+The script can be something like:
+
+```
+#!/bin/sh
+
+cat > /etc/conjur/ssl/client.pem <<EOF
+#{cert-content}
+EOF
+
+chmod 644 /etc/conjur/ssl/client.pem
+exit
+```
+
+This approach has the following benefits:
+1. Short URL. Certificate is in the script. (avoid issues raised from option 2)
+1. Deterministic operation. You know exactly when it is done, because the connection is closed once it reaches exit.
+1. Custom error handling. You can get any error you wish for, handle it properly, and send it back.
+1. Versatility. Script can be customized later if needed to handle more issues in this area.
+1. Simple implementation
+
+This solution is implemented in [this PR](https://github.com/cyberark/conjur/pull/1789).
+
```

Comment:
I actually don't understand these solutions in relation to the problem

The problem seems to me to be:
- The server sends a 200 to the client once the kube exec command is off, without waiting for a response to see if the command actually succeeds

Could we frame the solutions in relation to the problem? I'd be interested to know, for each potential solution:
- Does the server still return 200 to the client? At which point? If not, what alternate response codes does it send, and in which scenarios?
- Does the server wait to know if the kube exec / cert injection succeeds before returning a response to the client?

I would like to understand better the practical implications of each solution - what impact does it have on the server operations? What impact does it have on the client? It would be great if we could be crystal clear about those pros/cons at this stage, because we'll have to deal with those practical impacts once we get to implementation.

Reply:
Well ,the solution is separated into 2, unrelated, parts. The first one is to properly inject the certificate, in a way that we will log the failure if it occurred. The second one is to decide whether we wait for the cert to be injected before returning a response to the "inject_client_cert" request or not. The first part is answered in this section. The second part is answered in the following section - `Inject Client Cert response code`.

So in conclusion, this solution will answer both questions.

Reply:
I was actually suggesting it would be interesting to see how _each_ proposed solution would address the two parts of this problem. I do see you add more detail to that below for the solution that appears to be the winner, but the doc is missing context about how the other solutions fall short.

Reply:
i don't think that it is needed as the way that we are injecting the cert is independent to the question if we wait for the cert to be injected before returning the response. 

I re-arranged this section a bit to make it clearer.